### DISCUSSION TITLE
Unable to build a S3 client

### DISCUSSION THREAD

**Discussion by quaaanLee** [aws/aws-sdk-cpp#2805] [Q&A] [answered]
When I build my S3 client, I meet the complain like this:

```
aws-sdk-cpp/crt/aws-crt-cpp/crt/s2n/crypto/s2n_ktls_crypto.h:55:48: error: field ‘aes_gcm_256’ has incomplete type
   55 |         s2n_ktls_crypto_info_tls12_aes_gcm_256 aes_gcm_256;
```

while I find that it is defined right before in this header file:
```
#if defined(S2N_KTLS_SUPPORTED)
    #include <linux/tls.h>

    typedef struct tls12_crypto_info_aes_gcm_128 s2n_ktls_crypto_info_tls12_aes_gcm_128;
    typedef struct tls12_crypto_info_aes_gcm_256 s2n_ktls_crypto_info_tls12_aes_gcm_256;
#else
    #define TLS_1_2_VERSION 0

    #define TLS_CIPHER_AES_GCM_128 0
    typedef struct s2n_ktls_crypto_info_stub s2n_ktls_crypto_info_tls12_aes_gcm_128;
    #define TLS_CIPHER_AES_GCM_256 0
    typedef struct s2n_ktls_crypto_info_stub s2n_ktls_crypto_info_tls12_aes_gcm_256;
#endif
```

I use compile command like this:
```
cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=./awssdk -DBUILD_ONLY="s3" -DBUILD_SHARED_LIBS=0;

make -j16  // I meet problem when i execute this
```

Has anybody meet the same problem?  
Help me please, thanks a lot

**Comment by quaaanLee** [comment]
I found that my kernel version is too old and the tls.h header file does not contain `TLS_CIPHER_AES_GCM_256` and `tls12_crypto_info_aes_gcm_256`， so how can I undefine  the  S2N_KTLS_SUPPORTED once, and make it work across the whole project?

**Reply by jmklix** [reply]
Can you provide more details about your environment/compiler version/sdk version?

I would suggest trying to build/test s2n-tls on it's own to resolve your issue in isolation. It looks like there are some tests that might help you. In [this](https://github.com/aws/s2n-tls/tree/main/.github) folder there are some tests that will let you know if S2N_KTLS_SUPPORTED is supported. You can see someone else run them on this [issue](https://github.com/aws/s2n-tls/issues/4280)
```
-- feature S2N_KTLS_SUPPORTED: FALSE
```

**Reply by jmklix** [reply, marked as answer]
This might be fixed with a recent update of our crt dependencies. Can you try updating to the latest version of this sdk and let us know if you have any problems building.
